### PR TITLE
[ML] improve bpe tokenization to use logic not regex

### PR DESCRIPTION
**Pull request by benwtrent**
BPE tokenization uses a fairly trivial regex pattern to tokenize its strings. But, this regex on larger string inputs can be slow. This commit rewrites the regex into conditional logic to increase inference speed for bpe tokenization (roberta currently).

Here are some micro-benchmarks:

```
Benchmark                       Mode  Cnt    Score   Error  Units
BpeRegexVsLogicBenchmark.logic  avgt    9   59.068 ± 0.778  us/op
BpeRegexVsLogicBenchmark.regex  avgt    9  133.627 ± 5.922  us/op
```

Using logic is around 2x times faster (this is JUST with the tokenization). But, now we also can copy the string fewer times (as the logic works on a charsequence). So the overall performance improvement for tokenization is probably even better.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+import java.util.Queue;
+
+import static java.lang.Character.isDigit;
+import static java.lang.Character.isLetter;
+import static java.lang.Character.isWhitespace;
+
+/**
+ * This is a char sequence tokenizer that satisfies the regex bpe regex for gpt-2
+ *      "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)|\\s+"
+ *
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java:21]: Please add to the comment that in plain English this boils down to one of:
1. An ASCII apostrophe followed by one of a handful of specific letter sequences
2. Optional space followed by an unbroken sequence of letters
3. Optional space followed by an unbroken sequence of numbers
4. Optional space followed by an unbroken sequence of characters that are neither letters, numbers nor whitespace
5. Non-empty sequence of whitespace

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+ * It is by no means thread safe can only can parse one sequence at a time
+ */
+public class BpeTokenReader {
+    private final CharSequence inputChars;
+    int count;
+    private final Queue<Integer> putBackChar = new LinkedList<>();
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java:27]: `ArrayDeque` is supposed to be faster than `LinkedList`.  Every time an item is added or removed to a linked list it needs to create or destroy the link structures.  But an array backed structure can continually reuse the same backing array and just adjust the pointers to the start and end of what's in use.

**Review by droberts195 — APPROVED** [review]
LGTM if you could just expand one comment.

I also made a suggestion about the putback queue.  I'm not sure if any performance benefit would be measurable in practice.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BpeTokenReader.java:27]: @droberts195 I actually tried an arraydeque and it was slower in my benchmark (no matter the preallocated array deque size). In typical usage, we only ever put back one char. I was surprised by this, but the benchmark doesn't lie.
